concordium-consensus/src/Concordium/GlobalState/Persistent/Instances.hs

```diff
@@ -89,39 +95,106 @@ data PersistentInstance = PersistentInstance {
+           addVersion v f (s, inst') = (putWord8 v <> s, f inst')
+
     store pinst = fst <$> storeUpdate pinst
     load = do
-        rparams <- load
-        rInterface <- load
-        pinstanceModel <- get
-        pinstanceAmount <- get
-        return $ do
-            pinstanceParameters <- rparams
-            pinstanceModuleInterface <- rInterface
-            pip <- loadBufferedRef pinstanceParameters
-            let pinstanceHash = makeInstanceHash pip pinstanceModel pinstanceAmount
-            return PersistentInstance{..}
+      if demoteProtocolVersion (protocolVersion @pv) <= P3 then do
+        loadIV0
+      else
+        getWord8 >>= \case
+          0 -> loadIV0
+          1 -> loadIV1
+          n -> error $ "Unsupported persistent instance version " ++ show n
+      where 
+        loadIV0 = do
+          rparams <- load
+          rInterface <- load
+          pinstanceModel <- get
+          pinstanceAmount <- get
+          return $ do
+              pinstanceParameters <- rparams
+              pinstanceModuleInterface <- rInterface
+              pip <- loadBufferedRef pinstanceParameters
+              let pinstanceHash = makeInstanceHash pip pinstanceModel pinstanceAmount
+              return $! PersistentInstanceV0 (PersistentInstanceV {..})
+        loadIV1 = do
+          rparams <- load
+          rInterface <- load
+          pinstanceModel <- get
+          pinstanceAmount <- get
+          return $ do
+              pinstanceParameters <- rparams
+              pinstanceModuleInterface <- rInterface
+              pip <- loadBufferedRef pinstanceParameters
+              let pinstanceHash = makeInstanceHash pip pinstanceModel pinstanceAmount
+              return $! PersistentInstanceV1 (PersistentInstanceV {..})
 
 -- This cacheable instance is a bit unusual. Caching instances requires us to have access
 -- to the modules so that we can share the module interfaces from different instances.
-instance MonadBlobStore m => Cacheable (ReaderT Modules m) PersistentInstance where
-    cache p@PersistentInstance{..} = do
+instance MonadBlobStore m => Cacheable (ReaderT Modules m) (PersistentInstance pv) where
+    cache (PersistentInstanceV0 p@PersistentInstanceV{..}) = do
+        modules <- ask
+        lift $! do
+            -- we only cache parameters and get the interface from the modules
+            -- table. The rest is already in memory at this point since the
+            -- fields are flat, i.e., without indirection via BufferedRef or
+            -- similar reference wrappers.
+            ips <- cache pinstanceParameters
+            params <- loadBufferedRef ips
+            let modref = pinstanceContractModule params
+            miface <- Modules.unsafeGetModuleReferenceV0 modref modules
+            case miface of
+              Nothing -> return (PersistentInstanceV0 p{pinstanceParameters = ips}) -- this case should never happen, but it is safe to do this.
```

Comment:
What would the implications be for the caller when `PersistentInstanceV0` gets returned without the module interface.

Reply:
Since this is only caching the only thing that'd happen is that there would be two references to the module, one from the module table, and one from the instance. This would lead to the same module being loaded in memory twic.e